gensim/models/word2vec.py

```diff
@@ -1144,9 +1152,43 @@ def __iter__(self):
                         i += self.max_sentence_length
 
 
+def _scan_vocab_worker(stream, progress_queue, max_vocab_size=None, trim_rule=None):
+    """Do an initial scan of all words appearing in stream.
+
+    Note: This function can not be Word2VecVocab's method because
+    of multiprocessing synchronization specifics in Python.
+    """
+    min_reduce = 1
+    vocab = defaultdict(int)
+    checked_string_types = 0
+    sentence_no = -1
+    total_words = 0
+    for sentence_no, sentence in enumerate(stream):
+        if not checked_string_types:
+            if isinstance(sentence, string_types):
+                log_msg = "Each 'sentences' item should be a list of words (usually unicode strings). " \
+                          "First item here is instead plain %s." % type(sentence)
+                progress_queue.put(log_msg)
+
+            checked_string_types += 1
+
+        for word in sentence:
+            vocab[word] += 1
+
+        if max_vocab_size and len(vocab) > max_vocab_size:
+            utils.prune_vocab(vocab, min_reduce, trim_rule=trim_rule)
+            min_reduce += 1
```

Comment:
If I understand correctly, this step should also be applied to the final merged vocab, in addition to each individual vocab?

Reply:
`max_vocab_size` by itself doesn't guarantee that after vocabulary building its size will be `<= max_vocab_size`. This value is only used to throw away rare tokens during the building time. 

After "reduce" step (self.raw_vocab = `reduce(utils.merge_counts(....))`) another vocabulary method `prepare_vocab` is called, which prunes vocabulary appropriately.

Reply:
The `max_vocab_size` is a crude scan-time cap, but for a while it was the only cap available, and it does have the effect in prior code (since all vocabulary was discovered via the single-threaded process) that the final vocabulary could not possibly be larger than than its value. 

I don't see a way the later `prepare_vocab()` would apply that same cap (unless the user also specifies the newer, more precise `max_final_vocab` parameter). So the concern here is that a previously-reliable invariant – you'd never wind up with a vocabulary larger than `max_vocab_size` – may no longer apply.

Reply:
Yeah, I meant exactly `max_final_vocab` parameters. But it's true that if `max_final_vocab` isn't passed, the property `len(vocab) <= max_vocab_size` may not hold. 

Is it okay to prune the vocab after merging by sorting and selecting top `max_vocab_size` words by their frequency?

Reply:
@gojomo Actually, current single stream code can't guarantee `len(vocab) <= max_vocab_size` after vocabulary building. I think there is a bug in the code (https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/models/word2vec.py#L1185)

Say, you have only one sentence `a a a b b b b` and you set `max_vocab_size=1`. After only one iteration in scan_vocab, vocabulary will be the same `{"a": 3, "b": 4}` (pruning is called like this `utils.prune_vocab({"a": 3, "b": 4}, min_reduce=1) => {"a": 3, "b": 4}`).

Reply:
Yes, it's OK. Don't forget the main point of this parameter is not to blow up the RAM, it's there for a reason (applies even more to multistream).

Reply:
@persiyanov Yes, that looks like a bug, though one that would be vanishingly rare with more typical parameters (& Zipfian frequencies). I happen to really dislike this mechanism, because it's prone to misunderstanding - someone who, for example, would be happy with a 100,000 word vocabulary and sets `max_vocab_size=100000` might wind up with a vocab of far less if one of its draconian, escalating prunes (which often take out 90% of all words) happens just before the end-of-scan. But it'd be wise to either retain its (usual) hard-cap effect (which some may be relying upon), or better document that any true hard-cap should be enforced through the other parameter.

@piskvorky & @persiyanov In the current code `max_vocab_size` is enforcing any RAM-cap at all in the multistream case, is it? I'm not sure it has a straightforward interpretation for multi-stream, either - neither giving each stream its own `max_vocab_size` dict-size cap, nor giving each stream `max_vocab_size / num_streams` as a cap, does something intuitive/sensible.

Reply:
The objective is to limit the use of memory, so "total RAM used" would be the ideal parameter (ala Bounter). Barring that, a global `max_vocab_size` is a reasonable approximation.

Reply:
@piskvorky But, note that as currently written in this PR, `max_vocab_size` may be a single global parameter, but is only enforced locally in each stream-reader. So actual potential max-number-of-vocab-entries is `num_readers * max_vocab_size` – which is not likely what a user expects, especially if familiar with its previous single-thread interpretation. And, if any single reader hits the cap locally, it will be pruning its local counts only - other even smaller counts, of the exact same token in other threads, will survive – making any final tallies even more arbitrarily incorrect than in the single-thread application.

Reply:
Should definitely be `max_vocab / workers`, to keep the same semantics (~ same max memory footprint).

Reply:
@piskvorky fixed